tools/loader/Loader.cpp

```diff
@@ -379,7 +382,26 @@ void Loader::runInference(PlaceholderBindings &bindings, size_t batchSize) {
     timer.startTimer();
   }
   for (unsigned i = 0; i < iterationsOpt; i++) {
-    EE_.run(bindings);
```

Comment:
whoa the single run() method turned into the code below?  We should introduce a convenience method to wrap all of this.

Reply:
I think this is much more complicated because we have to create a promise, a future, an error to move the produced error into, etc. We could make this much simpler by splitting the `runInference` function into two parts:

```
void Loader::runInferences(std::unique_ptr<ExecutionContext> context) {
  assert(!emittingBundle() &&
         "No inference is performed in the bundle generation mode.");
  if (timeOpt) {
    timer_ = llvm::Timer timer("Infer", "Infer");
    timer_.startTimer();
  }

   auto iterations = iterationsOpt;
   runOneInference(std::move(context), --iterations);
}

void runOneInference(std::unique_ptr<ExecutionContext> context, size_t iterations) {
    hostManager->runNetwork(modelPathOpt[0], std::move(context), [this, iterations] 
        (runtime::RunIdentifierTy, llvm::Error err, std::unique_ptr<ExecutionContext> context) {
             EXIT_ON_ERR(std::move(runErr));
             if (iterations) {
                runOneInference(std::move(context), --iterations);
             } else {
                 if (timeOpt) {
                     timer_.stopTimer();
                     llvm::outs() << llvm::formatv("Wall time per item (s): {0:f4}\n",
                                               timer_.getTotalTime().getWallTime() /
                                               iterationsOpt / batchSize_);
                }
           }
      });
}
```

If you pull the timer code out into helper methods on the Loader class this is quite small.

Reply:
It's even simpler if you don't want to force the inferences to run sequentially and can just queue em all up in the HostManager and wait for `iterationsOpt` results.